### DISCUSSION TITLE
Pass List from Python to Rust without cloning elements

### DISCUSSION THREAD

**Discussion by cswinter** [PyO3/pyo3#2037] [Questions] [answered]
Is there a good way to receive a Python `List` without the elements getting cloned? (Context: [entity-neural-network/ragged-buffer:src/monomorphs/f32.rs#71](https://github.com/entity-neural-network/ragged-buffer/blob/7daee7f092396d3d057290fac06f94016573a6de/src/monomorphs/f32.rs#L71))

I've tried receiving a list and extracting each element, but the `extract` doesn't like returning a borrow:

```rust
    #[classmethod]
    fn cat(_cls: &PyType, buffers: &PyList, dim: usize) -> PyResult<Self> {
        let mut bufs = Vec::new();
        for buf in buffers.iter() {
            bufs.push(&buf.extract::<&RaggedBufferF32>()?.0);
        }
        Ok(RaggedBufferF32(RaggedBuffer::cat(&bufs, dim)?))
    }
```

**Comment by birkenfeld** [comment, marked as answer]
You can extract a `&PyCell<RaggedBufferF32>` and `borrow` it to get a `&RaggedBufferF32`.

**Reply by birkenfeld** [reply]
Even easier, you can extract `PyRef<RaggedBufferF32>` directly (which is what you would get from `borrow`).

**Reply by cswinter** [reply]
Nice, thank you! With `PyRef`, you don't even need the manual extract anymore:

```
    #[classmethod]
    fn cat(_cls: &PyType, buffers: Vec<PyRef<RaggedBufferF32>>, dim: usize) -> PyResult<Self> {
        Ok(RaggedBufferF32(RaggedBuffer::cat(
            &buffers.iter().map(|b| &b.0).collect::<Vec<_>>(),
            dim,
        )?))
    }
```

**Reply by birkenfeld** [reply]
Indeed, although now you're allocating one more intermediate Vec. Of course, this might not really matter.

**Comment by cswinter** [comment]
Yes, it doesn't matter too much in my case. With the current interface, you do need the second `Vec` as a backing array for the `PyRef` objects, but at least one vec alloc could be eliminated by changing the `RaggedBuffer::cat` to accepting `PyRef<RaggedBufferF32>` or `AsRef<RaggedBufferF32` instead of borrows.
